### PR TITLE
POL-201 AzureAD -> Flexera CMP Group Sync

### PR DESCRIPTION
**Pull request by srpomeroy**
### Description

Syncs Groups and their users from AzureAD to Flexera CMP

### Issues Resolved

n/a

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

**Comment by srpomeroy** [comment]
Internal ticket reference POL-21

File: operational/azure/azuread_group_sync/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+- *Email addresses of the recipients you wish to notify* - A list of email addresses to notify
+- *Prefix to filter groups on* - Uses the `startswith()` odata query to filter AzureAD groups based on a prefix
+  - Example: `FLEXERA_CMP_`
+- *Default Phone Number for New Users* - Use this phone number if the user doesn't have one set in AzureAD.
+  - The policy looks for a value in the first item of `businessPhones` attribute of the AzureAD user account.
+- *Default Company Name for New Users* - Use this Company Name if the user doesn't have one set in AzureAD
+  - The policy looks for a value in the `companyName` attribute of the AzureAD user account.
+- *Identity Provider Href* - The Href for the Identity Provider to associate to new users
+  - Can be retrieved by copying the link to edit your SSO: ht&#8203;tps://us-3.rightscale.com/global/enterprises/54321/edit_sso?identity_provider_id=**123** and grabbing the ID value at the end, or via the [RightScale Cloud Management API](https://reference.rightscale.com/api1.5/resources/ResourceIdentityProviders.html#index)
```
Comment (cdwilhelm) [inline operational/azure/azuread_group_sync/README.md]: should you use this link?

File: operational/azure/azuread_group_sync/azuread_group_sync.pt [hunk 2]
```diff
@@ -0,0 +1,754 @@
...
+    version: "1.0",
+    provider: "Azure",
+    service: "AzureAD"
+)
+
+###############################################################################
+# Permissions
+###############################################################################
+
+permission "perm_read_creds" do
```
Comment (cdwilhelm) [inline operational/azure/azuread_group_sync/azuread_group_sync.pt]: permissions block should be removed.

File: operational/azure/azuread_group_sync/azuread_group_sync.pt [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
+rs_pt_ver 20180301
+type "policy"
+short_description "Synchronizes AzureAD Groups to Flexera Governance Groups. \n See the [README](https://github.com/rightscale/policy_templates/tree/master/operational/azure/azuread_group_sync) and [docs.rightscale.com/policies](https://docs.rightscale.com/policies/) to learn more."
+long_description ""
+category "Operational"
+severity "low"
+tenancy "single"
+default_frequency "hourly"
+info(
+    version: "1.0",
```
Comment (cdwilhelm) [inline operational/azure/azuread_group_sync/azuread_group_sync.pt]: all cred mgr policies should start at version 2.0

File: operational/azure/azuread_group_sync/azuread_group_sync.pt [hunk 4]
```diff
@@ -0,0 +1,754 @@
...
+###############################################################################
+# Policy
+###############################################################################
+
+policy "policy_azuread_group_sync" do
+  validate $ds_combined_data do
+    summary_template "{{ len data.groupsToUpdate }} AzureAD Groups to Synchronize"
+    detail_template <<-EOS
+## Groups that Will Be Synchronized from AzureAD
+| Name  | Href  | AzureAD Object Id | AzureAD User Count | CMP User Count |
```
Comment (cdwilhelm) [inline operational/azure/azuread_group_sync/azuread_group_sync.pt]: sort by name?

File: operational/azure/azuread_group_sync/README.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+### Credential configuration
+
+For administrators [creating and managing credentials](https://docs.rightscale.com/policies/users/guides/credential_management.html) to use with this policy, the following information is needed:
+
+Provider tag value to match this policy: `azure_rm`
+
+Required permissions in the provider:
+
+- Tenant > Azure Active Directory Graph > Directory.Read.All (Application with Admin Consent)
+
```
[the remark this answers is not in the document]
Comment (srpomeroy) [reply operational/azure/azuread_group_sync/README.md]: It does use policy creds.

File: operational/azure/azuread_group_sync/README.md (hunk 1 above)
Comment (srpomeroy) [reply operational/azure/azuread_group_sync/README.md]: It's "fake" but I can replace it with a token.

File: operational/azure/azuread_group_sync/azuread_group_sync.pt (hunk 4 above)
Comment (srpomeroy) [reply operational/azure/azuread_group_sync/azuread_group_sync.pt]: I believe the AzureAD API returns results sorted. Are you asking that I forcibly sort it in the datasource just to be sure?

Comment (srpomeroy) [reply operational/azure/azuread_group_sync/azuread_group_sync.pt]: Added a sort to the groups to update and create.

Comment (cdwilhelm) [reply operational/azure/azuread_group_sync/azuread_group_sync.pt]: I didn't review the data, as long as the data is sorted in a logical way that's all we needed.  most people have been forcing with _sortBy()

**Comment by cdwilhelm** [comment]
@srpomeroy  are you done with all your changes?

**Comment by srpomeroy** [comment]
I am now @cdwilhelm.

**Pull request merged**